### PR TITLE
Switch AsyncCompletion to new pooled CompletionItemList based APIs

### PR DESCRIPTION
**Pull request by genlu**
Address [AB#1475048](https://devdiv.visualstudio.com/0bdbc590-a062-4c3f-b0f6-9383f67865ee/_workitems/edit/1475048)

LOH allocation for `CompletionItem[]` and `CompletionItemWithHighlight[]` are gone. 
image

this will break our CI integration tests, so I queued DartLab integration test run with latest VS build which passed
https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=6272608&view=results

### CODE REVIEW

File: src/EditorFeatures/Core.Cocoa/Microsoft.CodeAnalysis.EditorFeatures.Cocoa.csproj [hunk 1]
```diff
@@ -52,7 +52,7 @@
          validation errors on workload dependencies), we are instead manually
          referencing what we need out of the macos SDK. -->
     <PackageDownload Include="Microsoft.macOS.Ref" Version="[$(MicrosoftmacOSRefVersion)]" />
-    <Reference Include="$(NuGetPackageRoot)microsoft.macos.ref\$(MicrosoftmacOSRefVersion)\ref\net6.0\Xamarin.Mac.dll" />
+    <Reference Include="$(NuGetPackageRoot)microsoft.macos.ref\$(MicrosoftmacOSRefVersion)\ref\net6.0\Microsoft.macOS.dll" />
```
Comment (davidwengier) [inline src/EditorFeatures/Core.Cocoa/Microsoft.CodeAnalysis.EditorFeatures.Cocoa.csproj:55]: @sandyarmstrong this is the same change as you're making in https://github.com/dotnet/roslyn/pull/61250

Is this a problem? Or do we need to wait for the 6.0.300 bump in VS Mac?

Comment (genlu) [reply src/EditorFeatures/Core.Cocoa/Microsoft.CodeAnalysis.EditorFeatures.Cocoa.csproj:55]: Ping @sandyarmstrong, could you please take a look at this and let me know if it's OK? Thanks

**Comment by genlu** [comment]
The referenced version of editor package in latest commit is pending insertion
https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_git/VS/pullrequest/403261

File: src/EditorFeatures/Core.Cocoa/Microsoft.CodeAnalysis.EditorFeatures.Cocoa.csproj (hunk 1 above)
Comment (sandyarmstrong) [reply src/EditorFeatures/Core.Cocoa/Microsoft.CodeAnalysis.EditorFeatures.Cocoa.csproj:55]: You should merge https://github.com/dotnet/roslyn/pull/61250 first. But then we will not be able to insert Roslyn with this change for at least a few days while we finish up the work to bump VSMac to use the 6.0.300 SDK and the newer macos workload.

File: src/EditorFeatures/Core/IntelliSense/AsyncCompletion/CompletionSource.cs [hunk 2]
```diff
@@ -1,4 +1,4 @@
-﻿// Licensed to the .NET Foundation under one or more agreements.
+// Licensed to the .NET Foundation under one or more agreements.
```
Comment (sharwell) [inline src/EditorFeatures/Core/IntelliSense/AsyncCompletion/CompletionSource.cs:1]: Was this change intentional? It appears the file in the repository already has the correct encoding.

File: src/EditorFeatures/Core/IntelliSense/AsyncCompletion/CompletionSource.cs [hunk 3]
```diff
@@ -346,24 +346,21 @@ internal sealed class CompletionSource : IAsyncExpandingCompletionSource
...
 
-                if (context2.Items.IsEmpty && context2.SuggestionItemOptions is null)
+                if (context2.ItemList.IsEmpty && context2.SuggestionItemOptions is null)
                     return context1;
 
-                using var _ = ArrayBuilder<VSCompletionItem>.GetInstance(context1.Items.Length + context2.Items.Length, out var itemsBuilder);
-                itemsBuilder.AddRange(context1.Items);
-                itemsBuilder.AddRange(context2.Items);
-
+                var completionList = session.CreateCompletionList(context1.ItemList.Concat(context2.ItemList));
```
[the remark this answers is not in the document]
Comment (genlu) [reply src/EditorFeatures/Core/IntelliSense/AsyncCompletion/CompletionSource.cs:357]: That'd would require some changes to be made to `CompletionList` in Editor first. I will do this as an improvement in a follow PR

File: src/EditorFeatures/Core/IntelliSense/AsyncCompletion/CompletionSource.cs (hunk 2 above)
Comment (genlu) [reply src/EditorFeatures/Core/IntelliSense/AsyncCompletion/CompletionSource.cs]: fixed

File: src/EditorFeatures/Core.Cocoa/Preview/PreviewPane.cs [hunk 4]
```diff
@@ -31,7 +31,7 @@ public PreviewPane(string? id, string? title, Uri? helpLink, string? helpLinkToo
             var titlePlaceholder = new NSStackView()
             {
                 Orientation = NSUserInterfaceLayoutOrientation.Vertical,
-                EdgeInsets = new NSEdgeInsets(5, 0, 5, 0),
+                EdgeInsets = new NSEdgeInsets(new(5), new(0), new(5), new(0)),
```
[the remark this answers is not in the document]
Comment (genlu) [reply src/EditorFeatures/Core.Cocoa/Preview/PreviewPane.cs]: fixed

File: eng/Versions.props [hunk 5]
```diff
@@ -25,21 +25,21 @@
     <MicrosoftVisualStudioExtensibilityTestingVersion>0.1.132-beta</MicrosoftVisualStudioExtensibilityTestingVersion>
     <!-- CodeStyleAnalyzerVersion should we updated together with version of dotnet-format in dotnet-tools.json -->
     <CodeStyleAnalyzerVersion>4.2.0-2.final</CodeStyleAnalyzerVersion>
-    <VisualStudioEditorPackagesVersion>17.3.37-preview</VisualStudioEditorPackagesVersion>
+    <VisualStudioEditorPackagesVersion>17.3.122-preview</VisualStudioEditorPackagesVersion>
```
[the remark this answers is not in the document]
Comment (olegtk) [reply eng/Versions.props:28]: yeah, 17.3.122-preview is fine. Inserted in both main and rel/d17.3 for Preview2 now.

**Review by olegtk — APPROVED** [review]

File: src/EditorFeatures/Core.Cocoa/Microsoft.CodeAnalysis.EditorFeatures.Cocoa.csproj (hunk 1 above)
Comment (sandyarmstrong) [reply src/EditorFeatures/Core.Cocoa/Microsoft.CodeAnalysis.EditorFeatures.Cocoa.csproj:55]: I merged https://github.com/dotnet/roslyn/pull/61250, we should be ready to consume it in VSMac in the next few days.

File: src/EditorFeatures/Core/IntelliSense/AsyncCompletion/CompletionSource.cs (hunk 3 above)
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/IntelliSense/AsyncCompletion/CompletionSource.cs:357]: is it not possible to do CreateCompletionList, and then add both arrays to it in two AddRange calls?

File: src/EditorFeatures/Core/IntelliSense/AsyncCompletion/CompletionSource.cs [hunk 6]
```diff
@@ -444,26 +442,27 @@ static VSCompletionContext CombineCompletionContext(VSCompletionContext context1
...
             var filterSet = new FilterSet();
-            using var _ = ArrayBuilder<VSCompletionItem>.GetInstance(completionList.ItemsList.Count, out var itemsBuilder);
-
-            foreach (var roslynItem in completionList.ItemsList)
-            {
-                cancellationToken.ThrowIfCancellationRequested();
-                var item = Convert(document, roslynItem, filterSet, triggerLocation);
-                itemsBuilder.Add(item);
-            }
+            var completionItemList = session.CreateCompletionList(ConvertItems());
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/IntelliSense/AsyncCompletion/CompletionSource.cs:445]:
```suggestion
            var completionItemList = session.CreateCompletionList(
                completionList.ItemsList.Select(i => Convert(document, i, filterSet, triggerLocation, cancellationToken));
```

File: src/EditorFeatures/Core/IntelliSense/AsyncCompletion/CompletionSource.cs [hunk 7]
```diff
@@ -444,26 +442,27 @@ static VSCompletionContext CombineCompletionContext(VSCompletionContext context1
...
+            return (new(completionItemList, suggestionItemOptions, selectionHint: AsyncCompletionData.InitialSelectionHint.SoftSelection, filters, isIncomplete: false, null), completionList);
+
+            IEnumerable<VSCompletionItem> ConvertItems()
+            {
+                foreach (var roslynItem in completionList.ItemsList)
+                {
+                    cancellationToken.ThrowIfCancellationRequested();
+                    yield return Convert(document, roslynItem, filterSet, triggerLocation);
+                }
+            }
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/IntelliSense/AsyncCompletion/CompletionSource.cs:465]:
```suggestion
```

File: src/EditorFeatures/Core/IntelliSense/AsyncCompletion/ItemManager.cs [hunk 8]
```diff
@@ -2,19 +2,24 @@
...
 using Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion.Data;
 using Roslyn.Utilities;
+
+using RoslynCompletionItem = Microsoft.CodeAnalysis.Completion.CompletionItem;
 using VSCompletionItem = Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion.Data.CompletionItem;
 
 namespace Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.AsyncCompletion
 {
-    internal sealed partial class ItemManager : IAsyncCompletionItemManager
+    internal sealed partial class ItemManager : IAsyncCompletionItemManager2
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/IntelliSense/AsyncCompletion/ItemManager.cs:22]: can you link me to where thsi API is defined?

Comment (genlu) [reply src/EditorFeatures/Core/IntelliSense/AsyncCompletion/ItemManager.cs:22]: https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_git/VS-Platform?path=/src/Editor/Language/Def/Language/AsyncCompletion/IAsyncCompletionItemManager2.cs&_a=contents&version=GBmain

File: src/VisualStudio/IntegrationTest/New.IntegrationTests/InProcess/EditorInProcess.cs [hunk 9]
```diff
@@ -227,11 +227,7 @@ public async Task<ClassificationSpan[]> GetLightBulbPreviewClassificationsAsync(
 
             if (!string.IsNullOrEmpty(menuText))
             {
-                if (activeSession.TryGetSuggestedActionSets(out var actionSets) != QuerySuggestedActionCompletionStatus.Completed)
-                {
-                    actionSets = Array.Empty<SuggestedActionSet>();
-                }
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/IntegrationTest/New.IntegrationTests/InProcess/EditorInProcess.cs:233]: why did this cahnge?

File: src/VisualStudio/IntegrationTest/New.IntegrationTests/InProcess/LightBulbHelper.cs [hunk 10]
```diff
@@ -81,7 +81,7 @@ public static async Task<IEnumerable<SuggestedActionSet>> WaitForItemsAsync(Test
             // Calling PopulateWithData ensures the underlying session will call SuggestedActionsUpdated at least once
             // with the latest data computed.  This is needed so that if the lightbulb computation is already complete
             // that we hear about the results.
-            asyncSession.PopulateWithData(overrideRequestedActionCategories: null, operationContext: null);
+            await asyncSession.PopulateWithDataAsync(overrideRequestedActionCategories: null, operationContext: null).ConfigureAwait(false);
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/IntegrationTest/New.IntegrationTests/InProcess/LightBulbHelper.cs:84]: why these changes?

File: src/VisualStudio/IntegrationTest/TestUtilities/InProcess/Editor_InProc.cs [hunk 11]
```diff
@@ -378,11 +376,7 @@ public ClassifiedToken[] GetLightbulbPreviewClassifications(string menuText)
 
             if (!string.IsNullOrEmpty(menuText))
             {
-                if (activeSession.TryGetSuggestedActionSets(out var actionSets) != QuerySuggestedActionCompletionStatus.Completed)
-                {
-                    actionSets = Array.Empty<SuggestedActionSet>();
-                }
-
+                var actionSets = await LightBulbHelper.WaitForItemsAsync(broker, view).ConfigureAwait(false);
                 var set = actionSets.SelectMany(s => s.Actions).FirstOrDefault(a => a.DisplayText == menuText);
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/IntegrationTest/TestUtilities/InProcess/Editor_InProc.cs:380]: why these changes?

File: src/VisualStudio/TestUtilities2/ProjectSystemShim/Framework/TestEnvironment.vb [hunk 12]
```diff
@@ -78,7 +78,7 @@ Namespace Microsoft.VisualStudio.LanguageServices.UnitTests.ProjectSystemShim.Fr
             ExportProvider = composition.ExportProviderFactory.CreateExportProvider()
             _workspace = ExportProvider.GetExportedValue(Of VisualStudioWorkspaceImpl)
             ThreadingContext = ExportProvider.GetExportedValue(Of IThreadingContext)()
-            Interop.WrapperPolicy.s_ComWrapperFactory = MockComWrapperFactory.Instance
+            Implementation.Interop.WrapperPolicy.s_ComWrapperFactory = MockComWrapperFactory.Instance
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/TestUtilities2/ProjectSystemShim/Framework/TestEnvironment.vb:81]: why these changes?

File: src/VisualStudio/VisualBasic/Impl/LanguageService/VisualBasicPackage.vb [hunk 13]
```diff
@@ -57,7 +57,7 @@ Namespace Microsoft.VisualStudio.LanguageServices.VisualBasic
         Public Sub New()
             MyBase.New()
 
-            _comAggregate = Interop.ComAggregate.CreateAggregatedObject(Me)
+            _comAggregate = Implementation.Interop.ComAggregate.CreateAggregatedObject(Me)
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/VisualBasic/Impl/LanguageService/VisualBasicPackage.vb:60]: why these changes?

File: src/EditorFeatures/Core/IntelliSense/AsyncCompletion/ItemManager.cs (hunk 8 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/IntelliSense/AsyncCompletion/ItemManager.cs:22]: so, tbh... this api feels super strange to me:

```c#
    public interface IAsyncCompletionItemManager2 : IAsyncCompletionItemManager
    {
        /// <summary>
        /// This method is first called before completion is about to appear.
        /// The result of this method will be used in subsequent invocations of <see cref="IAsyncCompletionItemManager.UpdateCompletionListAsync"/>
        /// <paramref name="session"/> tracks user user's input tracked with <see cref="IAsyncCompletionSession.ApplicableToSpan"/>.
        /// <paramref name="data"/> provides applicable <see cref="AsyncCompletionSessionDataSnapshot.Snapshot"/> and 
        /// </summary>
        /// <param name="session">The active <see cref="IAsyncCompletionSession"/>. See <see cref="IAsyncCompletionSession.TextView"/></param>
        /// <param name="data">Contains properties applicable at the time this method is invoked.</param>
        /// <param name="token">Cancellation token that may interrupt this operation</param>
        /// <returns>Sorted <see cref="CompletionList{CompletionItem}"/> that will be subsequently passed to <see cref="IAsyncCompletionItemManager.UpdateCompletionListAsync"/></returns>
        Task<CompletionList<CompletionItem>> SortCompletionItemListAsync(
            IAsyncCompletionSession session,
            AsyncCompletionSessionInitialDataSnapshot data,
            CancellationToken token);
    }
```

It feels (imo) more natural if the lang server (us) just exposed an appropriate comparer for items.  Then the platform owns pushign the data back and forth from the snapshot object to the final CompletionList they create.

Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/IntelliSense/AsyncCompletion/ItemManager.cs:22]: WDYT?

File: src/EditorFeatures/Core/IntelliSense/AsyncCompletion/ItemManager.cs [hunk 14]
```diff
@@ -33,19 +38,22 @@ internal ItemManager(RecentItemsManager recentItemsManager, IGlobalOptionService
...
-            var sortedItems = data.InitialList.OrderBy(CompletionItemData.GetOrAddDummyRoslynItem).ToImmutableArray();
             AsyncCompletionLogger.LogItemManagerSortTicksDataPoint((int)stopwatch.Elapsed.TotalMilliseconds);
-            return Task.FromResult(sortedItems);
+            return Task.FromResult(items);
+        }
+
+        public Task<CompletionList<VSCompletionItem>> SortCompletionItemListAsync(
+            IAsyncCompletionSession session,
+            AsyncCompletionSessionInitialDataSnapshot data,
+            CancellationToken cancellationToken)
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/IntelliSense/AsyncCompletion/ItemManager.cs:50]: looks like we lose the cancellation tokrn and don't take advantage of it when sorting.

**Comment by CyrusNajmabadi** [comment]
Done with pass :)

File: src/EditorFeatures/Core/IntelliSense/AsyncCompletion/ItemManager.cs [hunk 15]
```diff
@@ -105,12 +110,51 @@ internal ItemManager(RecentItemsManager recentItemsManager, IGlobalOptionService
...
+                    AsyncCompletionLogger.LogSessionContainsTargetTypeFilter();
+            }
+
+            var items = new SegmentedList<VSCompletionItem>(data.InitialItemList);
+            items.Sort(VSItemComparer.Instance);
+
+            return items;
+        }
+
+        private sealed class VSItemComparer : IComparer<VSCompletionItem>
```
[the remark this answers is not in the document]
Comment (genlu) [reply src/EditorFeatures/Core/IntelliSense/AsyncCompletion/ItemManager.cs:139]: Yea, sounds good to me, I will bring this up with Editor team.

File: src/EditorFeatures/Core/IntelliSense/AsyncCompletion/ItemManager.cs (hunk 8 above)
Comment (genlu) [reply src/EditorFeatures/Core/IntelliSense/AsyncCompletion/ItemManager.cs:22]: Yup, and Editor can handle the allocations within its own boundary. I will bring this up with them

File: src/EditorFeatures/Core.Cocoa/Preview/PreviewPane.cs [hunk 16]
```diff
@@ -62,32 +63,32 @@ public PreviewPane(string? id, string? title, Uri? helpLink, string? helpLinkToo
...
                 this.titleField.AttributedStringValue = attributedStringTitle;
                 // We do this separately, because the title sometimes isn't there (i.e. no diagnostics ID)
                 // and we want the preview to stretch to the top
-                NSLayoutConstraint.Create(view, NSLayoutAttribute.Top, NSLayoutRelation.Equal, titlePlaceholder, NSLayoutAttribute.Bottom, 1, 0).Active = true;
+                NSLayoutConstraint.Create(view, NSLayoutAttribute.Top, NSLayoutRelation.Equal, titlePlaceholder, NSLayoutAttribute.Bottom, new NFloat(1), new NFloat(0)).Active = true;
             }
             else
             {
-                NSLayoutConstraint.Create(view, NSLayoutAttribute.Top, NSLayoutRelation.Equal, this, NSLayoutAttribute.Top, 1, 0).Active = true;
+                NSLayoutConstraint.Create(view, NSLayoutAttribute.Top, NSLayoutRelation.Equal, this, NSLayoutAttribute.Top, new NFloat(1), new NFloat(0)).Active = true;
```
Comment (sandyarmstrong) [inline src/EditorFeatures/Core.Cocoa/Preview/PreviewPane.cs:91]: None of the changes in this file should be necessary.

Comment (genlu) [reply src/EditorFeatures/Core.Cocoa/Preview/PreviewPane.cs]: You are right, reverted

File: eng/Versions.props (hunk 5 above)
[the remark this answers is not in the document]
Comment (genlu) [reply eng/Versions.props:28]: Sure

File: src/EditorFeatures/Core.Cocoa/Preview/PreviewPane.cs (hunk 16 above)
Comment (sandyarmstrong) [reply src/EditorFeatures/Core.Cocoa/Preview/PreviewPane.cs]: Thanks!

File: src/EditorFeatures/Core/IntelliSense/AsyncCompletion/CompletionSource.cs (hunk 3 above)
Comment (genlu) [reply src/EditorFeatures/Core/IntelliSense/AsyncCompletion/CompletionSource.cs:357]: https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_search?action=contents&text=pooledreadonlylist&type=code&lp=code-Project&filters=ProjectFilters%7BDevDiv%7DRepositoryFilters%7BVS-Platform%7D&pageSize=25&result=DefaultCollection/DevDiv/VS-Platform/GBmain//src/Editor/Core/Def/PooledObjects/PooledReadOnlyList.cs

Comment (genlu) [reply src/EditorFeatures/Core/IntelliSense/AsyncCompletion/CompletionSource.cs:357]: btw, I'm gonna propose a changeto Editor to address yours and Sam's concern.

File: src/EditorFeatures/Core/IntelliSense/AsyncCompletion/CompletionSource.cs [hunk 17]
```diff
@@ -444,26 +442,27 @@ static VSCompletionContext CombineCompletionContext(VSCompletionContext context1
...
             if (completionList.SuggestionModeItem is null)
-                return (new(items, suggestionItemOptions: null, selectionHint: AsyncCompletionData.InitialSelectionHint.RegularSelection, filters), completionList);
+                return (new(completionItemList, suggestionItemOptions: null, selectionHint: AsyncCompletionData.InitialSelectionHint.RegularSelection, filters, isIncomplete: false, null), completionList);
 
             var suggestionItemOptions = new AsyncCompletionData.SuggestionItemOptions(
                 completionList.SuggestionModeItem.DisplayText,
                 completionList.SuggestionModeItem.Properties.TryGetValue(CommonCompletionItem.DescriptionProperty, out var description) ? description : string.Empty);
 
-            return (new(items, suggestionItemOptions, selectionHint: AsyncCompletionData.InitialSelectionHint.SoftSelection, filters), completionList);
+            return (new(completionItemList, suggestionItemOptions, selectionHint: AsyncCompletionData.InitialSelectionHint.SoftSelection, filters, isIncomplete: false, null), completionList);
```
[the remark this answers is not in the document]
Comment (genlu) [reply src/EditorFeatures/Core/IntelliSense/AsyncCompletion/CompletionSource.cs:457]: Sure, I can clean up those usage in this and related documents in another PR

File: src/EditorFeatures/Core/IntelliSense/AsyncCompletion/ItemManager.cs (hunk 8 above)
Comment (olegtk) [reply src/EditorFeatures/Core/IntelliSense/AsyncCompletion/ItemManager.cs:21]: @AmadeusW

**Comment by genlu** [comment]
@CyrusNajmabadi PTAL, thanks!

**Comment by CyrusNajmabadi** [comment]
Integration tests seem to be totally broken.  :(

File: src/EditorFeatures/Core/IntelliSense/AsyncCompletion/CompletionSource.cs (hunk 3 above)
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/IntelliSense/AsyncCompletion/CompletionSource.cs:357]: which .Concat is this?  is it a single alloc that just streams over both lists?  or is it an O(N) alloc that copies everything into a new list?

File: src/EditorFeatures/Core/IntelliSense/AsyncCompletion/ItemManager.cs (hunk 8 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/IntelliSense/AsyncCompletion/ItemManager.cs:21]: @genlu are we waiting on this first?  what's the plan here?

File: src/EditorFeatures/Core/IntelliSense/AsyncCompletion/ItemManager.cs [hunk 18]
```diff
@@ -84,19 +110,16 @@ internal ItemManager(RecentItemsManager recentItemsManager, IGlobalOptionService
...
                         var (expandedContext, _) = await task.ConfigureAwait(false);
-                        if (expandedContext.Items.Length > 0)
+                        if (expandedContext.ItemList.Count > 0)
                         {
                             // Here we rely on the implementation detail of `CompletionItem.CompareTo`, which always put expand items after regular ones.
-                            var itemsBuilder = ImmutableArray.CreateBuilder<VSCompletionItem>(expandedContext.Items.Length + data.InitialSortedList.Length);
-                            itemsBuilder.AddRange(data.InitialSortedList);
-                            itemsBuilder.AddRange(expandedContext.Items);
-                            var combinedList = itemsBuilder.MoveToImmutable();
+                            var combinedItemList = session.CreateCompletionList(data.InitialSortedItemList.Concat(expandedContext.ItemList));
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/IntelliSense/AsyncCompletion/ItemManager.cs:116]: same question about this concat.  is it a single-alloc? or an O(n) copy?

File: src/EditorFeatures/Core/IntelliSense/AsyncCompletion/ItemManager.cs (hunk 15 above)
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/IntelliSense/AsyncCompletion/ItemManager.cs:139]: what's the result of this talk?

File: src/VisualStudio/IntegrationTest/New.IntegrationTests/InProcess/EditorInProcess.cs (hunk 9 above)
Comment (CyrusNajmabadi) [reply src/VisualStudio/IntegrationTest/New.IntegrationTests/InProcess/EditorInProcess.cs:233]: what is their recommendation that we use instead of TryGetSuggestedActionSets?

**Comment by CyrusNajmabadi** [comment]
CompletionList is now:

```c#
    public sealed class CompletionList<T> : IReadOnlyList<T>, IDisposable
```

Who is expected to dispose this.

Note: calling this CompletionList was *very* confusing as i thought that meant it was a list of completoin items.  When really it's just a list of anything, it's just used in completion.

**Comment by genlu** [comment]
> Who is expected to dispose this.

IT's associated to a async completion session, and Editor will free them after the session is done (dismissed/committed/etc)

File: src/EditorFeatures/Core/IntelliSense/AsyncCompletion/CompletionSource.cs (hunk 3 above)
Comment (genlu) [reply src/EditorFeatures/Core/IntelliSense/AsyncCompletion/CompletionSource.cs:357]: I think this is the one streams over both lists w/o copying (`System.Linq.Enumerable.Concat`)

**Comment by olegtk** [comment]
I agree CompletionList<T> is somewhat confusing name. It reflects existing state of completion API where completion items are passed as ImmutableArray<CompletionItem> and also ImmutableArray<CompletionItemWithHighlight> and CompletionItem and CompletionItemWithHighlight types do not have common supertype...
The only way to create  CompletionList<T> is through IAsyncCompletionSession though, so that somewhat makes it a bit clearer when used.

File: src/EditorFeatures/Core/IntelliSense/AsyncCompletion/ItemManager.cs (hunk 8 above)
Comment (genlu) [reply src/EditorFeatures/Core/IntelliSense/AsyncCompletion/ItemManager.cs:21]: No, we are not waiting on this, since this is following the same pattern of an existing API `IAsyncCompletionItemManager.SortCompletionItemListAsync`

File: src/EditorFeatures/Core/IntelliSense/AsyncCompletion/CompletionSource.cs (hunk 3 above)
Comment (olegtk) [reply src/EditorFeatures/Core/IntelliSense/AsyncCompletion/CompletionSource.cs:357]: please file a bug to track this addition, @genlu

File: src/EditorFeatures/Core/IntelliSense/AsyncCompletion/ItemManager.cs (hunk 8 above)
Comment (olegtk) [reply src/EditorFeatures/Core/IntelliSense/AsyncCompletion/ItemManager.cs:21]: Yeah, let's not block on this. It's pretty fundamental part of existing public completion API. The only difference of IAsyncCompletionItemManager2 over IAsyncCompletionItemManager is an ability to return pooled CompletionList\<CompletionItem>. It simplifies switching off ImmutableArray without changing much of existing functionality.
@AmadeusW, would be reasonable to allow passing (or injecting through MEF) a comparer instead of implementing this method?

File: src/EditorFeatures/Core/IntelliSense/AsyncCompletion/ItemManager.cs (hunk 15 above)
[the remark this answers is not in the document]
Comment (olegtk) [reply src/EditorFeatures/Core/IntelliSense/AsyncCompletion/ItemManager.cs:139]: @AmadeusW

**Comment by CyrusNajmabadi** [comment]
I'm pretty ok with teh code.  But i would like some sort of integration test pass to validate that this isn't going to break things for some unforseen reason.  Thanks!

File: src/EditorFeatures/Core/IntelliSense/AsyncCompletion/ItemManager.cs (hunk 8 above)
Comment (genlu) [reply src/EditorFeatures/Core/IntelliSense/AsyncCompletion/ItemManager.cs:21]: Given that `IAsyncCompletionItemManager2` is new, I think we can simply replace the method with a property `IComparer<CompletionItem> CompletiomItemComparer`.

Comment (genlu) [reply src/EditorFeatures/Core/IntelliSense/AsyncCompletion/ItemManager.cs:21]:
```cs
    public interface IAsyncCompletionItemManager2 : IAsyncCompletionItemManager
    {
        IComparer<CompletionItem> CompletiomItemComparer { get; }

        [Obsolete]
        Task<CompletionList<CompletionItem>> SortCompletionItemListAsync(
            IAsyncCompletionSession session,
            AsyncCompletionSessionInitialDataSnapshot data,
            CancellationToken token);
    }
```

Comment (olegtk) [reply src/EditorFeatures/Core/IntelliSense/AsyncCompletion/ItemManager.cs:21]: looks reasonable to me, but I'd need @AmadeusW to make a call here.

**Review by CyrusNajmabadi — APPROVED** [review]
